app/src/main/java/io/apicurio/registry/rules/RulesServiceImpl.java

```diff
@@ -45,37 +48,59 @@
     @Inject
     RuleExecutorFactory factory;
 
+    @Inject
+    RulesProperties rulesProperties;
+
     /**
      * @see io.apicurio.registry.rules.RulesService#applyRules(java.lang.String, io.apicurio.registry.types.ArtifactType, ContentHandle, io.apicurio.registry.rules.RuleApplicationType)
      */
     @Override
     public void applyRules(String artifactId, ArtifactType artifactType, ContentHandle artifactContent,
-                           RuleApplicationType ruleApplicationType) throws RuleViolationException {
-        boolean useGlobalRules = false;
+                          RuleApplicationType ruleApplicationType) throws RuleViolationException {
         @SuppressWarnings("unchecked")
         List<RuleType> rules = Collections.EMPTY_LIST;
         if (ruleApplicationType == RuleApplicationType.UPDATE) {
             rules = storage.getArtifactRules(artifactId);
         }
-        if (rules.isEmpty()) {
-            rules = storage.getGlobalRules();
-            useGlobalRules = true;
-        }
-        if (rules.isEmpty()) {
-            return;
-        }
         ContentHandle currentArtifactContent = null;
         if (ruleApplicationType == RuleApplicationType.UPDATE) {
             StoredArtifact currentArtifact = storage.getArtifact(artifactId);
             currentArtifactContent = currentArtifact.getContent();
         }
-        for (RuleType ruleType : rules) {
-            RuleConfigurationDto configurationDto = useGlobalRules ?
-                                                    storage.getGlobalRule(ruleType) : storage.getArtifactRule(artifactId, ruleType);
-            RuleExecutor executor = factory.createExecutor(ruleType);
-            RuleContext context = new RuleContext(artifactId, artifactType, configurationDto.getConfiguration(),
-                                                  currentArtifactContent, artifactContent);
-            executor.execute(context);
+
+        applyGlobalAndArtifactRules(artifactId, artifactType, currentArtifactContent, artifactContent, rules);
+    }
+
+    private void applyGlobalAndArtifactRules(String artifactId, ArtifactType artifactType, ContentHandle currentArtifactContent, ContentHandle updatedArtifactContent, List<RuleType> artifactRules) {
+
+        boolean useGlobalRules = false;
+        if (artifactRules.isEmpty()) {
+            artifactRules = storage.getGlobalRules();
+            useGlobalRules = true;
+        }
+
+        List<RuleType> filteredDefaultGlobalRules = rulesProperties.getFilteredDefaultGlobalRules(artifactRules);
+        if (!filteredDefaultGlobalRules.isEmpty()) {
+            // We have default global rules (with any storage rules filtered out) plus rules from storage, so concatenate the 2 lists
+            artifactRules = Stream.concat(artifactRules.stream(), filteredDefaultGlobalRules.stream())
+                .collect(Collectors.toList());
+        }

```

Comment:
I think this should be wrapped in a check for `useGlobalRules`.  Basically, if there are rules defined at the artifact level, we always use those - we never merge the global rules (configured **or** default) with the artifact rules.

Reply:
The logic is that configured artifact rules always override globals... the idea was that the user needs a way to say "enable validation but **not** compatibility" even when the compatibility rule has been configured globally.  However, now that I write this out, there is no way for the user to say "disable all rules" at the artifact level, because that would look the same as "no artifact rule configuration exists".  :( :(

Reply:
I'm open to solutions. :)

Reply:
Ping @Apicurio/developers

Reply:
The Validity rule can be given a `NONE` value - if we add a `NONE` for the Compatibility rule (and handle it in the checking code) then a user could override any global rules for a specific artifact. Does that work?

Thanks for clarifying the precedence of rules - I have refactored that code to hopefully clarify things.
Still working on why the tests are failing..

Reply:
Yeah I think using the `NONE` configuration option is the only solution I can think of as well.  I'll open a ticket to add that option to the compat rule.